22 crear rutas index y show para book

paul-ot opened the pull request:
Trello Card: https://trello.com/c/hsTWqbSf/22-crear-rutas-index-y-show-para-book

- Added active_model_serializers.  …
- Created serializer for Book model.
- Added new controllers for api (versioned) '/api/v1/...' to keep it as it appears in Apiary.
- Created versioned routes for the api.

paul-ot committed:
Added api controllers and routes for book index and show

paul-ot committed:
- Added active_model_serializers.
- Created serializer for Book model.
- Added new controllers for api (versioned) '/api/v1/...' to keep it as it appears in Apiary.
- Created versioned routes for the api.

mtejedorwolox commented on app/controllers/api/v1/base_controller.rb:
```diff
@@ -0,0 +1,4 @@
+class Api::V1::BaseController < ApplicationController
```
preferimos la sintaxis modular
```
module Api
  module V1
    class BaseController
      #
    end
  end
end
```

holywyvern replied:
Usamos la sintaxis modular, por un problema del autoloader de rails, rails 6 recien va a permitir arreglar esto, o es una de las promesas del nuevo autoloader.
El problema es que cuando escribis Api::V1::BaseController, no sabe si Api y V1 son modulos o clases.

paul-ot replied:
Ok, cambiado!

mtejedorwolox commented on app/controllers/api/v1/base_controller.rb:
```diff
@@ -0,0 +1,4 @@
+class Api::V1::BaseController < ApplicationController
+  before_action :authenticate_user!
```
no pondria el authenticate en un controller base ya que no da lugar a tener endpoints publicos, tambien le cambiaria el nombre de BaseController a ApiController que es mas descriptivo al contexto, base podria ser base de cualquier cosa

ApplicationController -> padre de controllers que tienen vistas (ej: active admin)
ApiController -> padre de controllers para json requests

mtejedorwolox replied:
Tambien prefiero que el ApiController este fuera del scope de Api::V1, pero no esta mal igual asi

paul-ot replied:
Arreglado todo, moví auth de ApiController a Api::V2::BooksController, también moví ApiController a /api/, fuera del scope v1/

mtejedorwolox commented on app/controllers/api/v1/books_controller.rb:
```diff
@@ -0,0 +1,19 @@
+class Api::V1::BooksController < Api::V1::BaseController
```
mismo, sintaxis modular
```
module Api
  module V1
    class BooksController < ApiController
```

paul-ot replied:
Listo.

mtejedorwolox commented on app/controllers/api/v1/books_controller.rb:
```diff
@@ -0,0 +1,19 @@
+class Api::V1::BooksController < Api::V1::BaseController
+  before_action :load_book, only: :show
```
un `before_action` pierde sentido si es para un solo EP

paul-ot replied:
Listo.

mtejedorwolox commented on app/controllers/api/v1/books_controller.rb:
```diff
@@ -0,0 +1,19 @@
+  def index
+    books = Book.all
+
+    return render json: books, each_serializer: BookSerializer
```
``
```suggestion
    render json: Book.all
```
el return es implicito, y el serializer tambien por tener el mismo nombre de clase

paul-ot replied:
Impresionante, cambiado!

mtejedorwolox commented on app/controllers/api/v1/books_controller.rb:
```diff
@@ -0,0 +1,19 @@
+  end
+
+  def show
+    return head 404 unless @book
```
```suggestion
    book = Book.find(params[:id])
    render json: book
```
lo mismo, el return esta implicito, tambien el serializer
find por default busca por id, y tira una excepcion ActiveRecord::RecordNotFound si no encuentra el documento

ya que usar un find para un show es lo mas comun, la logica de error al no encontrarlo la dejaria en el ApiController del cual hereda

```
rescue_from ActiveRecord::RecordNotFound, with: :not_found
# ...
private

def not_found
  render json: { error: "record not found with id #{params[:id]}" }, status: :not_found
end
# ...
```

paul-ot replied:
Toda la razón del mundo!

mtejedorwolox commented on config/routes.rb:
```diff
@@ -1,5 +1,10 @@
-  # devise_for :users
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+  namespace :api do
```
podes usar https://github.com/bploetz/versionist para esto
```
api_version(module: 'api/v1', path: { value: 'api/v1' }) do
  resources :books, only: [:index, :show]
end
```

paul-ot replied:
Implementado!

mtejedorwolox commented on config/routes.rb:
```diff
@@ -1,5 +1,10 @@
 Rails.application.routes.draw do
   mount_devise_token_auth_for 'User', at: 'auth'
```
aca prefiero hacer
`mount_devise_token_auth_for 'User', at: 'api/v1/auth'`
es solo una preferencia, para que los endpoints de autenticacion esten tambien bajo el namespace api/v1

paul-ot replied:
Implementado!

paul-ot replied:
me tomé el atrevimiento de sugerir un cambio, puse la ruta de auth de la siguiente manera:

```
Rails.application.routes.draw do
  api_version(module: 'api/v1', path: { value: 'api/v1' }) do
    mount_devise_token_auth_for 'User', at: 'auth'
    resources :books, only: [:index, :show]
  end
end
```

mtejedorwolox replied:
imagine que pensarias en hacer eso, y me encantaria que se pudiera hacer, pero DTA empieza a hacer cosas raras y fallan los endpoints cuando lo pones en un scope, conseguimos el mismo resultado dejandolo afuera con at: 'api/v1/auth'

paul-ot committed:
Changes:
- Changed Api controller definitions to a modular syntaxis.
- It has been moved auth to the BooksController, from ApiController.
- BaseController has been renamed to ApiController, and moved to the base api/ directory.
- Before action, in BooksController has been removed, and implemented #find instead of #find_by_id method.
- Removed explicit returns in BooksController.
- It has been removed the explicit serializer definitions.
- It has been added 'rescue_from' in ApiController, to catch ActiveRecord::RecordNotFound exceptions, raised by BooksController.show
- It has been installed and implemented the gem 'versionist', to define the api routes.
- It has been moved the 'mount_devise_token_auth_for' definition inside the 'api/v1/' scope.

paul-ot commented:
Listos los cambios, cuando puedan verifiquen nuevamente.
Gracias!
@mtejedorwolox @holywyvern

paul-ot committed:
Cambios solicitados.

paul-ot committed:
Changes

- Added Devise::Test::ControllerHelpers to the spec/rails_helper file.
- It has been added the BooksController class, to test index and show methods.
- I've also added the `config.action_mailer.default_url_options` config to the test environment configuration file.
- Added shared context file for authenticated user.

paul-ot committed:
Added new serializers for users and rents models.
New controller to allow us to create a rent or list rents of a specific user.
Added routes for the above controller.

mtejedorwolox commented on app/controllers/api/api_controller.rb:
```diff
@@ -0,0 +1,12 @@
+module Api
+  class ApiController < ApplicationController
```
no me convence api/api_controller, entiendo que es un controller que existe en el contexto de solo api, pero es el padre de todo, yo lo pondria en el mismo nivel que el application controller
 @holywyvern ??

holywyvern replied:
En este caso, no me hace tanto ruido porque hay un unico api controller, si lo quiere mover, no lo veo mal. Si lo mueve a API como hay una unica version, tampoco me molesta

paul-ot replied:
ApiController movido a la altura de ApplicationController.

mtejedorwolox commented on config/routes.rb:
```diff
@@ -1,5 +1,6 @@
-  # devise_for :users
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  api_version(module: 'api/v1', path: { value: 'api/v1' }) do
+    mount_devise_token_auth_for 'User', at: 'auth'
```
lo que comente antes de DTA :smile:

paul-ot replied:
Solucionado.

paul-ot committed:
- It has been moved the ApiController to the root controllers directory
- It has been moved the auth routes outside the api/v1 scope.

paul-ot commented:
listos los cambios, y arreglados conflictos

	ambos agregados:       .rubocop.yml
	ambos agregados:       app/controllers/api/v1/books_controller.rb
	ambos modificados:     app/controllers/application_controller.rb
	ambos agregados:       app/models/book.rb
	ambos agregados:       app/models/rent.rb
	ambos agregados:       app/models/user.rb
	ambos agregados:       config/initializers/devise.rb
	ambos modificados:     config/routes.rb
	ambos agregados:       db/migrate/20190204131055_devise_create_users.rb
	ambos agregados:       db/migrate/20190204140743_add_devise_token_auth_fields_to_users.rb
	ambos agregados:       db/migrate/20190204145406_add_confirmable_and_lockable_fields_to_users.rb
	ambos agregados:       spec/factories/rents.rb

al hacer pull del branch.

paul-ot committed:
Añadidas gemas dotenv-rails y sidekiq.
También instalé Redis.
Creado mailer para rents y trigger en modelo Rent, para enviar email al dueño del rent.
Configurado ActiveJob to use sidekiq.
Template html para el email.

paul-ot committed:
Añadido modelo BookSuggestion, para permitirles a los usuarios sugerir un nuevo libro.
Añadida migración y factory.

paul-ot committed:
Movidos todos los valores de configuraciones de email a secrets.

paul-ot committed:
añadido lógica para contar los rents de un usuario, en el UserSerializer

paul-ot committed:
Fixed code offenses detected by Rubocop.

mtejedorwolox commented on app/controllers/api/v1/rents_controller.rb:
```diff
@@ -0,0 +1,26 @@
+module Api
+  module V1
+    class RentsController < Api::V1::BaseController
+      before_action :set_user, only: %i[index create]
```
remove this

mtejedorwolox commented on app/controllers/api/v1/rents_controller.rb:
```diff
@@ -0,0 +1,26 @@
+      before_action :set_user, only: %i[index create]
+
+      def index
+        render json: @user.rents.all
```
```suggestion
        render json: user.rents.all
```

mtejedorwolox commented on app/controllers/api/v1/rents_controller.rb:
```diff
@@ -0,0 +1,26 @@
+      end
+
+      def create
+        rent = @user.rents.create!(rent_params)
```
```suggestion
        rent = user.rents.create!(rent_params)
```

mtejedorwolox commented on app/controllers/api/v1/rents_controller.rb:
```diff
@@ -0,0 +1,26 @@
+        params.permit(:book_id, :start_date, :end_date)
+      end
+
+      def set_user
```
```suggestion
      def user
        @user ||= User.find(params[:user_id])
      end
```

mtejedorwolox commented on app/serializers/user_serializer.rb:
```diff
@@ -0,0 +1,20 @@
+  0
+end
+
+def unerad_notifications_count
```
unread

holywyvern commented on app/controllers/api/v1/books_controller.rb:
```diff
@@ -0,0 +1,16 @@
+      before_action :authenticate_user!
+
+      def index
+        render json: Book.all
```
Todos los endpoint de listado deben estar paginados, como indica una de las primeras cards.

holywyvern commented on app/views/rent_mailer/new_rent_email.html.erb:
```diff
@@ -0,0 +1,7 @@
+<h1>New Rent!</h1>
+
+<ul>
+  <li>Rent start date: <%= @rent.created_at %></li>
```
@mtejedorwolox vos decis que deberiamos de decir que use slim o le permitimos erb ?
Amerita en este caso?

paul-ot committed:
- Fixed attribute name in serializer.
- Modified rents controller to use method 'user' instead of 'set_user'.

paul-ot commented:
All requested changes were made.
@mtejedorwolox @holywyvern

holywyvern commented on app/controllers/api/v1/rents_controller.rb:
```diff
@@ -1,14 +1,12 @@
+    class RentsController < ApiController
       def index
-        render json: @user.rents.all
+        render json: user.rents.all
```
"Todos los endpoints deben estar paginados"
Recordá usar la gema wor_paginate para esto

paul-ot replied:
@holywyvern, ayer me comentó @mtejedorwolox que la paginación no se aplica en este proyecto, y ya que la card está como "archivada", no debería aplicar paginación.

holywyvern replied:
Esta bien, entonces no dije nada

paul-ot commented:
Graciasss!

El jue., 14 feb. 2019 a las 10:32, Ramiro Rojo (<notifications@github.com>)
escribió:

> *@holywyvern* approved this pull request.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/wolox-training/pm-rails/pull/9#pullrequestreview-203742734>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ah26-z5rWVWANRxQ4fqI2OJCCe8alHqyks5vNWVfgaJpZM4anTt9>
> .
>


-- 
Paul Marclay
+54.345.493.0600
